frontend/dockerfile/shell/lex.go

```diff
@@ -472,3 +490,93 @@ func BuildEnvs(env []string) map[string]string {
 
 	return envs
 }
+
+// convert a shell-like wildcard pattern (? is a single char, * either the
+// shortest or longest (greedy) string) to an equivalent regular expression
+func convertShellPatternToRegex(pattern string, greedy bool) (*regexp.Regexp, error) {
+	var s scanner.Scanner
+	s.Init(strings.NewReader(pattern))
+	out := strings.Builder{}
+	out.Grow(len(pattern) + 4)
+
+	// match only at the beginning of the string
+	out.WriteByte('^')
+
+	// default: non-greedy wildcards
+	starPattern := ".*?"
+	if greedy {
+		starPattern = ".*"
+	}
+
+	for tok := s.Next(); tok != scanner.EOF; tok = s.Next() {
+		switch tok {
+		case '*':
+			out.WriteString(starPattern)
+			continue
+		case '?':
+			out.WriteByte('.')
+			continue
+		case '\\':
+			// } as part of ${} needs to be escaped, but the escape isn't part
+			// of the pattern
+			if s.Peek() == '}' {
+				continue
+			}
+			out.WriteRune('\\')
+			tok = s.Next()
+			if tok != '*' && tok != '?' && tok != '\\' {
+				return nil, errors.Errorf("invalid escape '\\%c'", tok)
+			}
+		// regex characters that need to be escaped
+		// escaping closing is optional, but done for consistency
+		case '[', ']', '{', '}', '.', '+', '(', ')', '|', '^', '$':
+			out.WriteByte('\\')
+		}
+		out.WriteRune(tok)
+	}
+	return regexp.Compile(out.String())
+}
+
+func trimPrefix(word, value string, greedy bool) (string, error) {
+	re, err := convertShellPatternToRegex(word, greedy)
+	if err != nil {
+		return "", errors.Errorf("invalid pattern (%s) in substitution: %s", word, err)
+	}
+
+	idx := re.FindStringIndex(value)
+	if idx != nil {
+		value = value[idx[1]:]
+	}
+	return value, nil
+}
+
+// reverse without avoid reversing escapes, i.e. a\*c -> c\*a
+func reverseStringButNotEscapes(pattern string) string {
+	patternBuilder := strings.Builder{}
+	patternBuilder.Grow(len(pattern))
+	patternRunes := []rune(pattern)
+	for i := len(patternRunes) - 1; i >= 0; i-- {
+		tok := patternRunes[i]
+		if i > 0 && patternRunes[i-1] == '\\' {
+			patternBuilder.WriteByte('\\')
+			i--
+		}
+		patternBuilder.WriteRune(tok)
+	}
+
+	return patternBuilder.String()
+}
```

Comment:
`reversePattern` is likely a better name. I also suspect that this code won't handle all escapes correctly. In particular, I think it handles `\\` incorrectly. `\\a` is its own literal but I think this will change it to `\a\` instead of `a\\` because it won't understand that the second `\` belongs to the first `\` instead of the `a`.

I think `strings.Builder` also makes this harder to follow. In particular, you can likely just use `[]byte` or `[]rune` directly.

Reply:
At first I thought, this was an edge case that shouldn't happen, but then I found at least three situations where we would do something like this.

I renamed the function and changed it to write directly to a `[]rune` in the input order so `\\` is handled correctly.